pkg/controller/directvolumemigration/rsync.go

```diff
@@ -1690,3 +1724,710 @@ func (t *Task) deleteProgressReportingCRs(client k8sclient.Client) error {
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "dvm-rsync-",
+			Namespace:    req.namespace,
+			Labels:       labels,
+		},
+		Spec: corev1.PodSpec{
+			RestartPolicy: corev1.RestartPolicyNever,
+			Volumes:       volumes,
+			Containers:    containers,
+			NodeName:      req.nodeName,
+			SecurityContext: &corev1.PodSecurityContext{
+				SupplementalGroups: req.pvInfo.supplementalGroups,
+				FSGroup:            req.pvInfo.fsGroup,
+				SELinuxOptions:     req.pvInfo.seLinuxOptions,
+			},
+		},
+	}
+	return clientPod
+}
+
+func (t *Task) prepareRsyncPodRequirements(srcClient compat.Client) ([]rsyncClientPodRequirements, error) {
+	req := []rsyncClientPodRequirements{}
+	cluster, err := t.Owner.GetSourceCluster(t.Client)
+	if err != nil {
+		return req, liberr.Wrap(err)
+	}
+	t.Log.Info("Getting image for Rsync client Pods that will be created on source MigCluster")
+	transferImage, err := cluster.GetRsyncTransferImage(t.Client)
+	if err != nil {
+		return req, liberr.Wrap(err)
+	}
+	t.Log.Info("Getting [NS => PVCWithSecurityContext] mappings for PVCs to be migrated")
+	pvcMap, err := t.getfsGroupMapForNamespace()
+	if err != nil {
+		return req, liberr.Wrap(err)
+	}
+	t.Log.Info("Getting Rsync password for Rsync client Pods")
+	password, err := t.getRsyncPassword()
+	if err != nil {
+		return req, liberr.Wrap(err)
+	}
+	t.Log.Info("Getting [PVC => NodeName] mapping for PVCs to be migrated")
+	pvcNodeMap, err := t.getPVCNodeNameMap()
+	if err != nil {
+		return req, liberr.Wrap(err)
+	}
+	t.Log.Info("Getting limits and requests for Rsync client Pods")
+	limits, requests, err := t.getPodResourceLists(CLIENT_POD_CPU_LIMIT, CLIENT_POD_MEMORY_LIMIT, CLIENT_POD_CPU_REQUEST, CLIENT_POD_MEMORY_REQUEST)
+	if err != nil {
+		return req, liberr.Wrap(err)
+	}
+	isPrivileged, _ := isRsyncPrivileged(srcClient)
+	t.Log.Info(fmt.Sprintf("Rsync client Pods will be created with privileged=[%v]", isPrivileged))
+	for ns, vols := range pvcMap {
+		// Get stunnel svc IP
+		svc := corev1.Service{}
+		key := types.NamespacedName{Name: DirectVolumeMigrationRsyncTransferSvc, Namespace: ns}
+		err := srcClient.Get(context.TODO(), key, &svc)
+		if err != nil {
+			return req, err
+		}
+		ip := svc.Spec.ClusterIP
+		// Add PVC volume mounts
+		for _, vol := range vols {
+			rsyncOptions := t.getRsyncOptions()
+			if vol.verify {
+				rsyncOptions = append(rsyncOptions, "--checksum")
+			}
+			podRequirements := rsyncClientPodRequirements{
+				pvInfo:    vol,
+				namespace: ns,
+				image:     transferImage,
+				password:  password,
+				resourceReq: corev1.ResourceRequirements{
+					Limits:   limits,
+					Requests: requests,
+				},
+				privileged:   isPrivileged,
+				nodeName:     pvcNodeMap[ns+"/"+vol.name],
+				destIP:       ip,
+				rsyncOptions: rsyncOptions,
+			}
+			req = append(req, podRequirements)
+		}
+	}
+	return req, nil
+}
+
+func (t *Task) getRsyncOperationsContext() (compat.Client, []rsyncClientPodRequirements, error) {
```

Comment:
wonder if we can rename `getRsyncOperationsContext` to getRsyncOperationsClientAndRequirements?

*Context in the name can be understand as returning context.Context